File: python/hackerrank/juniper-2016/aliens_and_portals/aliens_and_portals.py
'''
https://www.hackerrank.com/contests/juniper-2016/challenges/aliens-and-portals

Aliens and Portals

You are an engineer who has been abducted by aliens from Mars. The king of Mars needs your help in building portals to travel from one planet to another.
The cost of building a portal between two planets is twice the distance between them. You are given the number of planets N and distances between the planets. You need to find the minimum cost of building portals between planets such that the king can travel from one planet to any of the remaining  N-1 planets by only using the portals built. Also on building the portals, you will have to answer Q queries. In each query you are required to find the actual distance traversed while travelling from one planet to another planet using the portals. Note that the distance between any two planets is distinct.

Note: If the distance between two planets isn't given then a direct portal can't be built between these two planets.

Input Format

The first line contains three integers N, M and Q. N denotes the number of planets and Q denotes the number of queries.
Following  M lines contain 3 integers u,v,dist.  u is source planet id, v is the destination planet id and dist is the distance between the two planets in light years.
Following Q lines contains two integers x and y. You have to determine the distance travelled while moving from planet x to planet y via the portals.

Output Format

The first line contains the minimum total cost of building portals.
Following  lines should contain the answer to each query.

Sample Input
4 5 3  
0 1 10  
0 2 6  
0 3 5  
1 3 15  
2 3 4  
0 1  
0 2  
0 3  
Sample Output
38  
10  
9  
5  
Explanation
0 - 1
| \ |   
2 - 3

MST
0 - 1
 \ 
2 - 3
In this case, 3 portal will be built. One between planet 0 and 1 ,another between planet 0 and 3 and another between planet 2 and 3.
Minimum cost of building the portals is 2*(10+5+4)=38.
For travelling from planet 0 to planet 1 the king would use the portal built between planet 0 and 1.
He would be traversing a distance equal to 10 light years.
For travelling from planet 0 to planet 2, he would have to use the portal built between 0 and 3 and the portal built between 2 and 3.
He would be traversing a distance equal to 9 light years.
For travelling from planet 0 to planet 3, the king would use the portal built between 0 and 3.He would be traversing a distance equal to 5 light years.
'''


# Node for DSU
class Node(object):
	def __init__(self, key):
		self.rank = 0
		self.parent = None
		self.key = key

# Disjoint-set Union
class DSU(object):
	def __init__(self):
		self.nodes_lookup = {}

	def make_set(self, item):
		n = self.nodes_lookup.get(item)
		if n is None:
			n = Node(item)
			self.nodes_lookup[item] = n
		return n


	def find(self, item):
		def find_(node):
			if node.parent == None:
				return node

			node.parent = find_(node.parent)
			return node.parent

		item = self.make_set(item)
		return find_(item)

	def union(self, a, b):
		lrep = self.find(a)
		rrep = self.find(b)

		if lrep == rrep:
			return
		
		if lrep.rank <= rrep.rank:
			if lrep.rank == rrep.rank:
				rrep.rank += 1
			lrep.parent = rrep
		else:
			rrep.parent = lrep


# Graph
class Graph(object):
	def __init__(self, n, m):
		self.nV = n
		self.nE = m
		# Either store list of edges or the adjacency list
		self.adjlist = [[] for _ in xrange(n)]
		self.edges = []

	def add_edge_to_adjlist(self, src, dst, dist):
		self.adjlist[src].append((dst, dist))
		self.adjlist[dst].append((src, dist))

	def add_edge_to_list(self, src, dst, dist):
		self.edges.append((src, dst, dist))

	# print the adjacency list or the list of edges whichever
	# underlying representation the graph object uses
	def print_(self):
		if not self.edges:
			for i,x in enumerate(self.adjlist, 0):
				print i, ':', x
		else:
			for e in self.edges:
				print e
			print  self.edges



	# Generate a Minimum Spanning Tree using Kruskal's algorithm
	def kruskalsMST(self):
		mst = Graph(self.nV, self.nV-1)
		dsu = DSU()

		num_edges = 0
		mst_weight = 0
		# sort edges by their distances
		for e in sorted(self.edges, key=lambda (u,v,d): d):
			if num_edges == self.nV - 1:
				break

			src, dst, dist = e
			if dsu.find(src) != dsu.find(dst):
				dsu.union(src, dst)
				mst.add_edge_to_adjlist(src, dst, dist)
				mst_weight += dist
				num_edges += 1

		# print the cost of building the portals
		# which is twice the distances between the MST edges
		# => 2 * weight of the MST = total cost
		print 2*mst_weight
		#mst.print_()
		return mst


	# Calculate distance between src -> dst using bfs
	def bfs_distance(self, src, dst):
		visited = [False] * self.nV
		q = [(src, 0)]
		while q:
			curr, dist = q.pop(0)
			if curr == dst:
				return dist

			for (v, vdist) in self.adjlist[curr]:
				if not visited[v]:
					q.append((v, dist+vdist))
					visited[v] = True

		# 'If' the graph is disconnected !!
		return -1

	# Calculate distance between src -> dst using dfs
	def dfs_distance(self, src, dst):
		def dfs(curr, curr_dist):
			if curr == dst:
				dist[0] = curr_dist
				return

			for (v, vdist) in self.adjlist[curr]:
				if not visited[v]:
					visited[v] = True
					dfs(v, curr_dist+vdist)
			
		visited = [False] * self.nV
		dist = [0]
		visited[src] = True
		dfs(src, 0)
		return dist[0]






def aliens_and_portals():
	n, m, q = map(int, raw_input().strip().split())
	g = Graph(n, m)

	while m > 0:
		src, dst, dist = map(int, raw_input().strip().split())
		g.add_edge_to_list(src, dst, dist)
		m -= 1

	#g.print_()
	mst = g.kruskalsMST()

	while q > 0:
		src, dst = map(int, raw_input().strip().split())
		dist = mst.bfs_distance(src, dst)

		# An MST should return the same distance between src->dst
		# using both bfs and dfs
		assert dist == mst.dfs_distance(src, dst)
		print dist
		q -= 1


if __name__ == '__main__':
	aliens_and_portals()

File: c++/hackerrank/juniper_2016/aliens_and_portals/aliens_and_portals.cpp
/**
https://www.hackerrank.com/contests/juniper-2016/challenges/aliens-and-portals

Aliens and Portals

You are an engineer who has been abducted by aliens from Mars. The king of Mars needs your help in building portals to travel from one planet to another.
The cost of building a portal between two planets is twice the distance between them. You are given the number of planets N and distances between the planets. You need to find the minimum cost of building portals between planets such that the king can travel from one planet to any of the remaining  N-1 planets by only using the portals built. Also on building the portals, you will have to answer Q queries. In each query you are required to find the actual distance traversed while travelling from one planet to another planet using the portals. Note that the distance between any two planets is distinct.

Note: If the distance between two planets isn't given then a direct portal can't be built between these two planets.

Input Format

The first line contains three integers N, M and Q. N denotes the number of planets and Q denotes the number of queries.
Following  M lines contain 3 integers u,v,dist.  u is source planet id, v is the destination planet id and dist is the distance between the two planets in light years.
Following Q lines contains two integers x and y. You have to determine the distance travelled while moving from planet x to planet y via the portals.

Output Format

The first line contains the minimum total cost of building portals.
Following  lines should contain the answer to each query.

Sample Input
4 5 3  
0 1 10  
0 2 6  
0 3 5  
1 3 15  
2 3 4  
0 1  
0 2  
0 3  
Sample Output
38  
10  
9  
5  
Explanation
0 - 1
| \ |   
2 - 3

MST
0 - 1
 \ 
2 - 3
In this case, 3 portal will be built. One between planet 0 and 1 ,another between planet 0 and 3 and another between planet 2 and 3.
Minimum cost of building the portals is 2*(10+5+4)=38.
For travelling from planet 0 to planet 1 the king would use the portal built between planet 0 and 1.
He would be traversing a distance equal to 10 light years.
For travelling from planet 0 to planet 2, he would have to use the portal built between 0 and 3 and the portal built between 2 and 3.
He would be traversing a distance equal to 9 light years.
For travelling from planet 0 to planet 3, the king would use the portal built between 0 and 3.He would be traversing a distance equal to 5 light years.

*/

#include <assert.h>
#include <vector>
#include <queue>
#include <iostream>
#include <utility>
#include <unordered_map>
#include <algorithm>

// Node for DSU
struct Node {
	int rank;
	int key;
	Node *parent;

	Node(int key) {
		this->key = key;
		this->rank = 0;
		this->parent = 0;
	}
};

// Disjoint set / union-find
class DisjointSet {
	std::unordered_map<int, Node*> nodes_lookup;

public:
	DisjointSet() {
		nodes_lookup = std::unordered_map<int, Node*>();
	}

	~DisjointSet() {
		for (auto node : nodes_lookup)
			delete node.second;
	}

	Node* make_set(int key) {
		if (nodes_lookup.find(key) == nodes_lookup.end()) {
			nodes_lookup[key] = new Node(key);
		}
		return nodes_lookup[key];
	}

	Node* find(Node *nodeitem) {
		if (!nodeitem->parent) {
			return nodeitem;
		}

		nodeitem->parent = this->find(nodeitem->parent);
		return nodeitem->parent;
	}

	Node* find(int item) {
		Node *n = make_set(item);
		return find(n);
	}


	void union_(int src, int dst) {
		Node *lrep, *rrep;

		lrep = find(src);
		rrep = find(dst);

		if (lrep == rrep)
			return;

		if (lrep->rank <= rrep->rank) {
			if (lrep->rank == rrep->rank)
				lrep->rank += 1;
			lrep->parent = rrep;
		} else {
			rrep->parent = lrep;
		}
	}
};


struct Edge {
	int src;
	int dst;
	int distance;

	Edge(int src, int dst, int distance) {
		this->src = src;
		this->dst = dst;
		this->distance = distance;
	}
};


class Graph {
	int nVertices;
	int nEdges;
	std::vector<std::pair<int, int>> *adjlist;
	std::vector<Edge> edges_list;

public:
	Graph(int nVertices, int nEdges) {
		this->nVertices = nVertices;
		this->nEdges = nEdges;
		this->adjlist = new std::vector<std::pair<int, int>>[this->nVertices];
		this->edges_list = std::vector<Edge>();
	}

	~Graph() {
		/** Free allocation for the array of vectors */
		delete[] this->adjlist;
	}


	/** Add a bi-directional edge from src->dst and dst-> src into the adjacency list */
	void add_edge_to_adjlist(int src, int dst, int distance) {
		this->adjlist[src].push_back(std::make_pair(dst, distance));
		this->adjlist[dst].push_back(std::make_pair(src, distance));
	}

	/** Add edge to edges list */
	void add_edge_to_list(int src, int dst, int distance) {
		this->edges_list.push_back(Edge(src, dst, distance));
	}


	/** Read graph edges' endpoints and link them */
	void read_graph()  {
		for (int i=0; i<this->nEdges; i++) {
			int src, dst, distance;

			std::cin >> src >> dst >> distance;
			add_edge_to_list(src, dst, distance);
		}
	}


	/** Print the adjacency list/edges list (whichever is used in the graph backend) of the graph */
	void print(void) {
		std::cout << "Graph: " << std::endl;
		if (this->edges_list.size() != 0) {
			for (Edge e : this->edges_list)
				std::cout << e.src << " " << e.dst << " " << e.distance << std::endl;

			return;
		}

		for (int i=0; i<this->nVertices; i++) {
			std::vector<std::pair<int,int>> &v = this->adjlist[i];
			std::cout << i << ": ";
			for (auto x : v)
				std::cout << "(" << x.first << "," << x.second << ") ";
			std::cout << std::endl;
		}
	}


	/** Generate a Minimum Spanning Tree using Kruskal's algorithm */
	Graph* kruskalsMST(void) {
		Graph *mst = new Graph(this->nVertices, this->nVertices-1);
		DisjointSet dsu;

		int num_edges = 0;
		int total_cost = 0;

		std::vector<Edge> sorted_edges = this->edges_list;
		auto EdgeComparator = [](Edge e1, Edge e2) {
			return (e1.distance < e2.distance);
		};
		std::sort(sorted_edges.begin(), sorted_edges.end(), EdgeComparator);

		for (Edge e : sorted_edges) {
			if (num_edges == (this->nVertices-1))
				break;
	
			if (dsu.find(e.src) != dsu.find(e.dst)) {
				dsu.union_(e.src, e.dst);
				mst->add_edge_to_adjlist(e.src, e.dst, e.distance);
				total_cost += e.distance;
				num_edges++;
			}
		}

		//mst->print();
		// total cost is 2*total weight of the MST
		std::cout<< 2*total_cost << std::endl;

		return mst;
	}

	/** Return the distance between src -> dst using BFS */
	int bfs_distance(int src, int dst) {
		std::queue <std::pair<int, int> > q;
		std::vector<bool> visited(this->nVertices, false);

		visited[src] = true;
		q.push(std::make_pair(src, 0));
		while (!q.empty()) {
			std::pair<int, int> p = q.front();
			q.pop();
			int curr = p.first, dist = p.second;

			if (dst == curr)
				return dist;

			for (auto e : this->adjlist[curr]) {
				int v = e.first;
				int vdist = e.second;
				if (!visited[v]) {
					visited[v] = true;
					q.push(std::make_pair(v, dist+vdist));
				}
			}

		}

		// Completed BFS for all reachable nodes,
		// couldn't find 'dst'
		return -1;
	}


	/** Recursive DFS traversal helper */
	void dfs(int curr, int curr_dist, int dst, int *distance, std::vector<bool> &visited) {
		if (curr == dst) {
			*distance = curr_dist;
			return;
		}

		for (auto e : this->adjlist[curr]) {
			int v = e.first;
			int vdist = e.second;
			if (!visited[v]) {
				visited[v] = true;
				dfs(v, curr_dist+vdist, dst, distance, visited);
			}
		}
	}

	/** Return the distance between src -> dst using DFS */
	int dfs_distance(int src, int dst) {
		std::vector<bool> visited(this->nVertices, false);

		visited[src] = true;
		int distance = 0;

		dfs(src, 0, dst, &distance, visited);

		return distance;
	}
};


int main(void)
{
	int n, m;
	int nQueries;

	std::cin >> n >> m >> nQueries;

	Graph g(n, m);;
	g.read_graph();
	//g.print();

	Graph *mst = g.kruskalsMST();

	while (nQueries--) {
		int src, dst, dist;
		std::cin >> src >> dst;
		dist = mst->bfs_distance(src, dst);

		// An MST should return the same distance between src->dst
		// using both bfs and dfs
		assert (mst->dfs_distance(src, dst) == dist);
		std::cout << dist << std::endl;
	}

	return 0;
}


